### ISSUE TITLE
`Hash::hash_slice` is not well-formed

### ISSUE THREAD

**Issue by apasel422** [rust-lang/rust#29426]
`foo.rs`:
```rust
use std::hash::{Hash, Hasher};

struct Foo<T: ?Sized>(T);

impl<T: ?Sized + Hash> Hash for Foo<T> {
    fn hash<H: Hasher>(&self, h: &mut H) {
        self.0.hash(h);
    }

    fn hash_slice<H: Hasher>(data: &[Self], h: &mut H) {
        unimplemented!()
    }
}

fn main() {}
```
Output from `rustc 1.5.0-nightly (c3db627cb 2015-10-26)`:
```
foo.rs:10:5: 12:6 warning: the trait `core::marker::Sized` is not implemented for the type `T` [E0277]
foo.rs:10     fn hash_slice<H: Hasher>(data: &[Self], h: &mut H) {
foo.rs:11         unimplemented!()
foo.rs:12     }
foo.rs:10:5: 12:6 help: run `rustc --explain E0277` to see a detailed explanation
foo.rs:10:5: 12:6 note: `T` does not have a constant size known at compile-time
foo.rs:10:5: 12:6 note: this warning results from recent bug fixes and clarifications; it will become a HARD ERROR in the next release. See RFC 1214 for details.
foo.rs:10     fn hash_slice<H: Hasher>(data: &[Self], h: &mut H) {
foo.rs:11         unimplemented!()
foo.rs:12     }
foo.rs:10:5: 12:6 note: required because it appears within the type `Foo<T>`
foo.rs:10:5: 12:6 note: slice and array elements must have `Sized` type
```
There should be a `where Self: Sized` clause on `hash_slice`.

**Comment by apasel422** [comment]
Never mind. Was looking at the wrong thing.
